README.md
# Barkup

[![Build Status](https://travis-ci.org/keighl/barkup.png?branch=master)](
https://travis-ci.org/keighl/barkup) [![Coverage Status](https://coveralls.io/repos/keighl/barkup/badge.svg?branch=master)](https://coveralls.io/r/keighl/barkup?branch=master)

[godoc.org/github.com/keighl/barkup](http://godoc.org/github.com/keighl/barkup)

Barkup is a library for backing things up. It provides tools for writing bare-bones backup programs in Go. The library is broken out into **exporters** and **storers**. Currently, those are:

**Exporters:** `MySQL` `Postgres` `RethinkDB`

**Storers:** `S3`

## Quick Example (mysql to s3)

Here's a go program that backups up a MySQL database (`Exporter`) to an S3 bucket (`Storer`) using barkup. The resulting executable is plopped on a server somewhere and scheduled to execute via CRON.

```go
package main

import "github.com/keighl/barkup"

func main() {

  // Configure a MySQL exporter
  mysql := &barkup.MySQL{
    Host: "localhost",
    Port: "3306",
    DB: "production_db",
    User: "root",
    Password: "cheese",
  }

  // Configure a S3 storer
  s3 := &barkup.S3{
    Region: "us-east-1",
    Bucket: "backups",
    AccessKey: "*************",
    ClientSecret: "**********************",
  }

  // Export the database, and send it to the
  // bucket in the `db_backups` folder
  err := mysql.Export().To("db_backups/", s3)
  if (err != nil) { panic(err) }
}
```

```
$ go build
```

```
@hourly /path/to/backup-program
```

## Exporters

Exporters provide a common interface for backing things up via the `Export()` method. It writes an export file to the local disk, and returns an `ExportResult` which can be passed on to a [storer](#storers), or to another location on the disk.

```go
// Exporter
mysql := &barkup.MySQL{...}

// Export Result
result := mysql.Export()
if (result.Error != nil) { panic(result.Error) }

// Send it to a directory path on a storer
err := result.To("backups/", storer)

// OR just move it somewhere on the local disk
err := result.To("~/backups/", nil)
```

---

### MySQL

The mysql exporter uses `mysqldump` to pump out a gzipped archive of your database. `mysqldump` must be installed on your system (it probably is if you're using mysql), and accessible to the user running the final program (again, it probabaly is).

**Usage**

```go
mysql := &barkup.MySQL{
  Host: "127.0.0.1",
  Port: "3306",
  DB: "XXXXX",
  User: "XXXXX",
  Password: "XXXXX",
  // Any extra mysqldump options
  Options: []string{"--skip-extended-insert"}
}

// Writes a file `./bu_DBNAME_TIMESTAMP.sql.tar.gz`
result := mysql.Export()

if (result.Error != nil) { panic(result.Error) }
```
---

### Postgres

The postgres exporter uses `pg_dump` to make a gzipped archive of your database. `pg_dump` must be installed on your system (it probably is if you're using postgres).

**Usage**

```go
postgres := &barkup.Postgres{
  Host: "127.0.0.1",
  Port: "5432",
  DB: "XXXXXXXX",

  // Not necessary if the program runs as an authorized pg user/role
  Username: "XXXXXXXX",

  // Any extra pg_dump options
  Options: []string{"--no-owner"},
}

// Writes a file `./bu_DBNAME_TIMESTAMP.sql.tar.gz`
result := postgres.Export()

if (result.Error != nil) { panic(result.Error) }
```

**Connection credentials**

You have two options for allowing barkup to connect to your DB.

Add a [`~/.pgpass`](http://www.postgresql.org/docs/9.3/static/libpq-pgpass.html) for account that will run the backup program.

Or, run the backup program from an authenticated user, like `postgres`:

```bash
$ sudo -i -u postgres
$ ./backup-program
```
---

### RethinkDB

The RethinkDB exporter uses `rethinkdb dump` to make a gzipped archive of your cluster. `rethinkdb` must be installed on your system.

**Usage**

```go
rethink := &barkup.RethinkDB{
  Name: "nightly",
  Connection: "0.0.0.0:28015",
  // You can specify specific databases and/or tables to dump (by default it dumps your whole cluster)
  Targets: []string{"site", "leads.contacts"},
}

// Writes a file `./bu_nightly_TIMESTAMP.tar.gz`
result := rethink.Export()
if (result.Error != nil) { panic(result.Error) }
```

## Storers

Storers take an `ExportResult` object and provide a common interface for moving a backup to someplace safe.

```go
// For chaining an ExportRestult
err := someExportResult.To("backups/", someStorer)

// OR
err := someStorer.Store(someExportResult, "backups/")
```

---

### S3

The S3 storer puts the exported file into a bucket at a specified directory. **Note,** you shouldn't use your global AWS credentials for this. Instead, [create bucket specific credentials via IAM.](http://blogs.aws.amazon.com/security/post/Tx3VRSWZ6B3SHAV/Writing-IAM-Policies-How-to-grant-access-to-an-Amazon-S3-bucket)

**Usage**

```go
s3 := &barkup.S3{
  Region: "us-east-1",
  Bucket: "backups",
  AccessKey: "XXXXXXXXXXXXX",
  ClientSecret: "XXXXXXXXXXXXXXXXXXXXX",
}

err := someExportResult.To("data/", s3)
```

**Region IDs**

* us-east-1
* us-west-1
* us-west-2
* eu-west-1
* ap-southeast-1
* ap-southeast-2
* ap-northeast-1
* sa-east-1



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 0.0.3 - 2015-01-30

* RethinkDB exports
* Refactored some methods
* Changed up test-stubbing technique for better code coverage

## 0.0.2 - 2014-12-07

* Postgres exports
* More documentation

## 0.0.1 - 2014-11-30

* Intitial version!
* MySQL exports
* AWS store


s3_test.go
package barkup

import (
  "testing"
  "launchpad.net/goamz/aws"
  "errors"
  "os"
)

func Test_S3_Store_Success(t *testing.T) {
  server := testServer(200, "", "application/json")
  aws.Regions["fake-region"] = aws.Region{
    Name: "fake-region",
    S3Endpoint: server.URL,
  }

  s := &S3{
    Region: "fake-region",
    Bucket: "cheese",
    AccessKey: "afjlsdfjkaf",
    ClientSecret: "adsfljdsahfl",
  }

  err := s.Store(&ExportResult{"test/test.txt", "text/plain", nil}, "test/")
  expect(t, err, nil)
}

func Test_S3_Store_Fail(t *testing.T) {
  server := testServer(500, "", "application/json")
  aws.Regions["fake-region"] = aws.Region{
    Name: "fake-region",
    S3Endpoint: server.URL,
  }

  s := &S3{
    Region: "fake-region",
    Bucket: "cheese",
    AccessKey: "afjlsdfjkaf",
    ClientSecret: "adsfljdsahfl",
  }

  _, _ = os.Create("test/test.txt")
  err := s.Store(&ExportResult{"test/test.txt", "text/plain", nil}, "test/")
  refute(t, err, nil)
}

func Test_S3_Store_ExportError(t *testing.T) {
  s := &S3{
    Region: "fake-region",
    Bucket: "cheese",
    AccessKey: "afjlsdfjkaf",
    ClientSecret: "adsfljdsahfl",
  }

  _, _ = os.Create("test/test.txt")
  err := s.Store(&ExportResult{"", "text/plain", errors.New("*****")}, "test/")
  refute(t, err, nil)
}

mysql_test.go
package barkup

import (
  "testing"
  "strings"
)

func Test_MySQL_Export_Pass(t *testing.T) {
  m := MySQL{
    Host: "localhost",
    Port: "3306",
    DB: "test",
    User: "root",
    Password: "cheese",
  }

  mysqlDumpCmd = "true"
  tarCmd = "true"

  result := m.Export()
  expect(t, result.Error, nil)
}

func Test_MySQL_Export_FailDump(t *testing.T) {
  m := MySQL{
    Host: "localhost",
    Port: "3306",
    DB: "test",
    User: "root",
    Password: "cheese",
  }

  mysqlDumpCmd = "false"
  tarCmd = "true"

  result := m.Export()
  refute(t, result.Error, nil)
}

func Test_MySQL_Export_FailTar(t *testing.T) {
  m := MySQL{
    Host: "localhost",
    Port: "3306",
    DB: "test",
    User: "root",
    Password: "cheese",
  }

  mysqlDumpCmd = "true"
  tarCmd = "false"

  result := m.Export()
  refute(t, result.Error, nil)
}

func Test_MySQL_optionsDump(t *testing.T) {
  m := MySQL{
    Host: "localhost",
    Port: "3306",
    DB: "test",
    User: "root",
    Password: "cheese",
    Options: []string{"--skip-extended-insert"},
  }

  options := strings.Join(m.dumpOptions(), " ")
  expect(t, strings.Contains(options, "-h"), true)
  expect(t, strings.Contains(options, "-P"), true)
  expect(t, strings.Contains(options, "-u"), true)
  expect(t, strings.Contains(options, "-p"), true)
  expect(t, strings.Contains(options, "--skip-extended-insert"), true)
  expect(t, strings.Contains(options, m.DB), true)
}



barkup_test.go
package barkup

import (
  "testing"
  "reflect"
  "net/http/httptest"
  "net/http"
  "fmt"
  "os"
  "errors"
)

func expect(t *testing.T, a interface{}, b interface{}) {
  if a != b {
    t.Errorf("Expected %v (type %v) - Got %v (type %v)", b, reflect.TypeOf(b), a, reflect.TypeOf(a))
  }
}

func refute(t *testing.T, a interface{}, b interface{}) {
  if a == b {
    t.Errorf("Did not expect %v (type %v) - Got %v (type %v)", b, reflect.TypeOf(b), a, reflect.TypeOf(a))
  }
}

func testServer(code int, body string, contentType string) *httptest.Server {
  return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    w.WriteHeader(code)
    w.Header().Set("Content-Type", contentType)
    fmt.Fprintln(w, body)
  }))
}

//////

func Test_ExportRestult_To_Move(t *testing.T) {
  file, _ := os.Create("to_mv_test")
  defer file.Close()

  e := ExportResult{"to_mv_test", "text/plain", nil}
  err := e.To("test/", nil)
  expect(t, err, nil)

  err = os.Remove("test/to_mv_test")
  expect(t, err, nil)
}

type StoreSuccessStory struct {}
func (x *StoreSuccessStory) Store(r *ExportResult, d string) (error) {
  return nil
}

type StoreFailureStory struct {}
func (x *StoreFailureStory) Store(r *ExportResult, d string) (error) {
  return errors.New("*****")
}

func Test_ExportRestult_To_Store(t *testing.T) {
  _, _ = os.Create("test/test.txt")
  e := &ExportResult{"test/test.txt", "text/plain", nil}
  err := e.To("test/", &StoreSuccessStory{})
  expect(t, err, nil)
}

func Test_ExportRestult_To_Store_Fail(t *testing.T) {
  _, _ = os.Create("test/test.txt")
  e := &ExportResult{"test/test.txt", "text/plain", nil}
  err := e.To("test/", &StoreFailureStory{})
  refute(t, err, nil)
}



s3.go
package barkup

import (
  "os"
  "bufio"
  "launchpad.net/goamz/aws"
  "launchpad.net/goamz/s3"
)

// S3 is a `Storer` interface that puts an ExportResult to the specified S3 bucket. Don't use your main AWS keys for this!! Create read-only keys using IAM
type S3 struct {
  // Available regions:
  // * us-east-1
  // * us-west-1
  // * us-west-2
  // * eu-west-1
  // * ap-southeast-1
  // * ap-southeast-2
  // * ap-northeast-1
  // * sa-east-1
  Region string
  // Name of the bucjet
  Bucket string
  // AWS S3 access key
  AccessKey string
  // AWS S3 secret
  ClientSecret string
}

// Puts an `ExportResult` struct to an S3 bucket within the specified directory
func (x *S3) Store(result *ExportResult, directory string) (error) {

  if (result.Error != nil) {
    return result.Error
  }

  file, err := os.Open(result.Path)
  if err != nil { return err }
  defer file.Close()

  buffy := bufio.NewReader(file)
  stat, err := file.Stat()
  if err != nil { return err }

  size := stat.Size()

  auth := aws.Auth{
    AccessKey: x.AccessKey,
    SecretKey: x.ClientSecret,
  }

  s := s3.New(auth, aws.Regions[x.Region])
  bucket := s.Bucket(x.Bucket)

  err = bucket.PutReader(directory + result.Filename(), buffy, size, result.MIME, s3.BucketOwnerFull)
  return err
}

mysql.go
package barkup

import (
  "fmt"
  "time"
  "os/exec"
  "os"
)

var tarCmd string = "tar"
var mysqlDumpCmd string = "mysqldump"

// MySQL is an `Exporter` interface that backs up a MySQL database via the `mysqldump` command
type MySQL struct {
  // DB Host (e.g. 127.0.0.1)
  Host string
  // DB Port (e.g. 3306)
  Port string
  // DB Name
  DB string
  // DB User
  User string
  // DB Password
  Password string
  // Extra mysqldump options
  // e.g []string{"--extended-insert"}
  Options []string
}

// Produces a `mysqldump` of the specified database, and creates a gzip compressed tarball archive.
func (x MySQL) Export() (*ExportResult) {
  result := &ExportResult{MIME: "application/x-tar"}

  dumpPath := fmt.Sprintf(`bu_%v_%v.sql`, x.DB, time.Now().Unix())

  options := append(x.dumpOptions(), fmt.Sprintf(`-r%v`, dumpPath))
  _, err := exec.Command(mysqlDumpCmd, options...).Output()
  result.Error = err
  if (result.Error != nil) { return result }

  result.Path = dumpPath+".tar.gz"
  _, err = exec.Command(tarCmd, "-czf", result.Path, dumpPath).Output()
  result.Error = err
  if (err != nil) { return result }
  os.Remove(dumpPath)

  return result
}

func (x MySQL) dumpOptions() []string {
  options := x.Options
  options = append(options, fmt.Sprintf(`-h%v`, x.Host))
  options = append(options, fmt.Sprintf(`-P%v`, x.Port))
  options = append(options, fmt.Sprintf(`-u%v`, x.User))
  options = append(options, fmt.Sprintf(`-p%v`, x.Password))
  options = append(options, x.DB)
  return options
}

barkup.go
package barkup

import (
  "os/exec"
  "path/filepath"
  "os"
)

//////////////

// The `Export()` method is expected to export "something" to a file and return a complete `ExportResult` struct (`Path`, `MIME`, `Error`). If any error occurs during it's work, it should set the error to the result's `Error` attribute
type Exporter interface {
  Export() (*ExportResult, error)
}

//////////////

// The `Store()` method takes an `ExportResult` and move it somewhere! To a cloud storage service, for instance...
type Storer interface {
  Store(result *ExportResult, directory string) (error)
}

//////////////

type ExportResult struct {
  // Path to exported file
  Path string
  // MIME type of the exported file (e.g. application/x-tar)
  MIME string
  // Any error that occured during `Export()`
  Error error
}

// Hands off an ExportResult to a `Storer` interface and invokes its Store() method. The directory argument is passed along too. If `store` is `nil`, the the method will simply move the export result to the specified directory (via the `mv` command)
func (x *ExportResult) To(directory string, store Storer) (error) {
  if (store == nil) {
    _, err := exec.Command("mv", x.Path, directory + x.Filename()).Output()
    return err
  }

  err := store.Store(x, directory)
  if (err != nil) { return err }

  err = os.Remove(x.Path)
  return err
}

// Returns the just filename component of the `Path` attribute
func (x ExportResult) Filename() (string) {
  _, filename := filepath.Split(x.Path)
  return filename
}




postgres.go
package barkup

import (
  "fmt"
  "time"
  "os/exec"
)

var pgDumpCmd string = "pg_dump"

// Postgres is an `Exporter` interface that backs up a Postgres database via the `pg_dump` command
type Postgres struct {
  // DB Host (e.g. 127.0.0.1)
  Host string
  // DB Port (e.g. 5432)
  Port string
  // DB Name
  DB string
  // Connection Username
  Username string
  // Extra pg_dump options
  // e.g []string{"--inserts"}
  Options []string
}

// Produces a `pg_dump` of the specified database, and creates a gzip compressed tarball archive.
func (x Postgres) Export() (*ExportResult) {
  result := &ExportResult{MIME: "application/x-tar"}
  result.Path = fmt.Sprintf(`bu_%v_%v.sql.tar.gz`, x.DB, time.Now().Unix())
  options := append(x.dumpOptions(), "-Fc", fmt.Sprintf(`-f%v`, result.Path))
  _, err := exec.Command(pgDumpCmd, options...).Output()
  result.Error = err
  return result
}

func (x Postgres) dumpOptions() []string {
  options := x.Options

  if x.DB != "" {
    options = append(options, fmt.Sprintf(`-d%v`, x.DB))
  }

  if x.Host != "" {
    options = append(options, fmt.Sprintf(`-h%v`, x.Host))
  }

  if x.Port != "" {
    options = append(options, fmt.Sprintf(`-p%v`, x.Port))
  }

  if x.Username != "" {
    options = append(options, fmt.Sprintf(`-U%v`, x.Username))
  }

  return options
}



rethinkdb.go
package barkup

import (
  "fmt"
  "time"
  "os/exec"

)

var rethinkCmd string = "rethinkdb"

// RethinkDB is an `Exporter` interface that backs up a RethinkDB cluster via the `rethinkdb dump` command
type RethinkDB struct {
  // Describe the dataset you're backing up. Used in filenaming
  // e.g. website
  Name string
  // HOST:PORT Connection (e.g. 127.0.0.1:28015)
  Connection string
  // Limit the dump to specific databases or tables in the cluster
  // e.g. `[]string{"main_website", "internal_website.users"}`
  Targets []string
  // Extra "rethinkdb dump" options
  // See `rethinkdb dump -h`
  // e.g  `[]string{"--authAUTHKEY"}`
  Options []string
}

// Produces a gzip compressed tarball archive of the rethink cluster (or targetted DBs/tables)
func (x RethinkDB) Export() (*ExportResult) {
  result := &ExportResult{MIME: "application/x-tar"}
  result.Path = fmt.Sprintf(`bu_%v_%v.tar.gz`, x.Name, time.Now().Unix())
  options := append(x.dumpOptions(), fmt.Sprintf(`-f%v`, result.Path))
  _, err := exec.Command(rethinkCmd, options...).Output()
  result.Error = err
  return result
}

func (x RethinkDB) dumpOptions() []string {
  options := append([]string{"dump"}, x.Options...)
  options = append(options, fmt.Sprintf(`-c%v`, x.Connection))
  for _, target := range x.Targets {
    options = append(options, fmt.Sprintf(`-e%v`, target))
  }
  return options
}

doc.go
/*
Barkup is a library for backing things up. It provides tools for writing bare-bones backup programs in Go. The library is broken out into exporters and storers. Currently, those are: MySQL, Postgres, S3

Quick Example

Here's a go program that backups up a MySQL database (Exporter) to an S3 bucket (Storer) using barkup. The resulting executable is plopped on a server somewhere and scheduled to execute via CRON.

    package main

    import "github.com/keighl/barkup"

    func main() {

      // Configure a MySQL exporter
      mysql := &barkup.MySQL{
        Host: "localhost",
        Port: "3306",
        DB: "production_db",
        User: "root",
        Password: "cheese",
      }

      // Configure a S3 storer
      s3 := &barkup.S3{
        Region: "us-east-1",
        Bucket: "backups",
        AccessKey: "*************",
        ClientSecret: "**********************",
      }

      // Export the database, and send it to the
      // bucket in the db_backups folder
      err := mysql.Export().To("db_backups/", s3)
      if (err != nil) { panic(err) }
    }

...

    $ go build

...

    @hourly /path/to/backup-program

Exporters

Exporters provide a common interface for backing things up via the Export() method. It writes an export file to the local disk, and returns an ExportResult which can be passed on to a storer, or to another location on the disk.

    // Exporter
    mysql := &barkup.MySQL{...}

    // Export Result
    result := mysql.Export()
    if (result.Error != nil) { panic(result.Error) }

    // Send it to a directory path on a storer
    err := result.To("backups/", storer)

    // OR just move it somewhere on the local disk
    err := result.To("~/backups/", nil)

MySQL

The mysql exporter uses mysqldump to pump out a gzipped archive of your database. mysqldump must be installed on your system (it probably is if you're using mysql), and accessible to the user running the final program (again, it probabaly is).

    mysql := &barkup.MySQL{
      Host: "127.0.0.1",
      Port: "3306",
      DB: "XXXXX",
      User: "XXXXX",
      Password: "XXXXX",
      // Any extra mysqldump options
      Options: []string{"--skip-extended-insert"}
    }

    // Writes a file ./bu_DBNAME_TIMESTAMP.sql.tar.gz
    result := mysql.Export()

    if (result.Error != nil) { panic(result.Error) }

Postgres

The postgres exporter uses pg_dump to make a gzipped archive of your database. pg_dump must be installed on your system (it probably is if you're using postgres).

    postgres := &barkup.Postgres{
      Host: "127.0.0.1",
      Port: "5432",
      DB: "XXXXXXXX",

      // Not necessary if the program runs as an authorized pg user/role
      Username: "XXXXXXXX",

      // Any extra pg_dump options
      Options: []string{"--no-owner"},
    }

    // Writes a file ./bu_DBNAME_TIMESTAMP.sql.tar.gz
    result := postgres.Export()

    if (result.Error != nil) { panic(result.Error) }

Connection credentials

You have two options for allowing barkup to connect to your DB. Add a [~/.pgpass](http://www.postgresql.org/docs/9.3/static/libpq-pgpass.html) for account that will run the backup program.

Or, run the backup program from an authenticated user, like postgres:

    $ sudo -i -u postgres
    $ ./backup-program

RethinkDB

The RethinkDB exporter uses `rethinkdb dump` to make a gzipped archive of your cluster. `rethinkdb` must be installed on your system.

Usage

    rethink := &barkup.RethinkDB{
      Name: "nightly",
      Connection: "0.0.0.0:28015",
      // You can specify specific databases and/or tables to dump (by default it dumps your whole cluster)
      Targets: []string{"site", "leads.contacts"},
    }

    // Writes a file `./bu_nightly_TIMESTAMP.tar.gz`
    result := rethink.Export()
    if (result.Error != nil) { panic(result.Error) }

Storers

Storers take an ExportResult object and provide a common interface for moving a backup to someplace safe.

    // For chaining an ExportRestult
    err := someExportResult.To("backups/", someStorer)

    // OR
    err := someStorer.Store(someExportResult, "backups/")

S3

The S3 storer puts the exported file into a bucket at a specified directory. Note, you shouldn't use your global AWS credentials for this. Instead, [create bucket specific credentials via IAM.](http://blogs.aws.amazon.com/security/post/Tx3VRSWZ6B3SHAV/Writing-IAM-Policies-How-to-grant-access-to-an-Amazon-S3-bucket)

    s3 := &barkup.S3{
      Region: "us-east-1",
      Bucket: "backups",
      AccessKey: "XXXXXXXXXXXXX",
      ClientSecret: "XXXXXXXXXXXXXXXXXXXXX",
    }

    err := someExportResult.To("data/", s3)

Region IDs

* us-east-1
* us-west-1
* us-west-2
* eu-west-1
* ap-southeast-1
* ap-southeast-2
* ap-northeast-1
* sa-east-1

*/
package barkup